File: lc_soln/0084_largest_rectangle_in_histogram/solution.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        result = 0
        n = len(heights)
        lefts = [-1] * n
        stack = []
        for i, h in enumerate(heights):
            while stack and heights[stack[-1]] >= h:
                stack.pop()
            if stack:
                lefts[i] = stack[-1]
            stack.append(i)

        stack = []
        rights = [n] * n
        for i, h in reversed(list(enumerate(heights))):
            while stack and heights[stack[-1]] >= h:
                stack.pop()
            if stack:
                rights[i] = stack[-1]
            stack.append(i)

        result = 0
        for i in range(n):
            result = max(result, heights[i] * (rights[i] - lefts[i] - 1))
        return result


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0084_largest_rectangle_in_histogram/solution.rs
struct Solution;

impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut lefts = vec![-1; heights.len()];
        let mut stack = vec![];
        for (i, &h) in heights.iter().enumerate() {
            while let Some(&j) = stack.last() && heights[j] >= h {
                stack.pop();
            }
            if let Some(&j) = stack.last() {
                lefts[i] = j as i32;
            }
            stack.push(i);
        }
        let mut stack = vec![];
        let mut rights = vec![heights.len() as i32; heights.len()];
        for (i, &h) in heights.iter().enumerate().rev() {
            while let Some(&j) = stack.last() && heights[j] >= h {
                stack.pop();
            }
            if let Some(&j) = stack.last() {
                rights[i] = j as i32;
            }
            stack.push(i);
        }
        let mut result = 0;
        for (i, &h) in heights.iter().enumerate() {
            result = result.max(h * (rights[i] - lefts[i] - 1));
        }
        result
    }
}

fn main() {
}
